src/Raven.Server/Commercial/LicenseManager.cs

```diff
@@ -121,40 +161,138 @@ public void Initialize(StorageEnvironment environment, TransactionContextPool co
 
             try
             {
-                _licenseStatus.Attributes = LicenseValidator.Validate(license, RSAParameters);
-                _licenseStatus.Error = false;
-                _licenseStatus.Message = null;
+                _recalculateLicenseLimitsSemsphore.Wait();
+
+                var licenseLimits = _serverStore.LoadLicenseLimits() ?? new LicenseLimits();
+                var coresByNode = licenseLimits.CoresByNode ?? 
+                    (licenseLimits.CoresByNode = new Dictionary<string, int>());
+
+                using (_serverStore.ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+                using (context.OpenReadTransaction())
+                {
+                    if (assignedNodeTag != null && assignedCores != null)
+                    {
+                        coresByNode[assignedNodeTag] = assignedCores.Value;
+                    }
+
+                    var allNodeTags = _serverStore.GetClusterTopology(context).AllNodes.Keys;
+                    var coresInUse = coresByNode.Sum(x => x.Value);
+                    if (allNodeTags.Count == coresByNode.Count &&
+                        coresInUse <= _licenseStatus.MaxCores)
+                    {
+                        return;
+                    }
+
+                    var nodesToRemove = coresByNode.Keys.Except(allNodeTags).ToList();
+                    foreach (var nodeToRemove in nodesToRemove)
+                    {
+                        coresByNode.Remove(nodeToRemove);
+                    }
+
+                    var missingNodesAssignment = new List<string>();
+                    foreach (var nodeTag in allNodeTags)
+                    {
+                        if (coresByNode.TryGetValue(nodeTag, out var _) == false)
+                            missingNodesAssignment.Add(nodeTag);
+                    }
+
+                    if (missingNodesAssignment.Count > 0)
+                    {
+                        var availableCores = _licenseStatus.MaxCores - coresInUse;
+                        var coresPerNode = Math.Max(1, availableCores / missingNodesAssignment.Count);
+                        foreach (var nodeTag in missingNodesAssignment)
+                        {
+                            coresByNode[nodeTag] = coresPerNode;
+                        }
+                    }
+
+                    coresInUse = coresByNode.Sum(x => x.Value);
+                    if (coresInUse > _licenseStatus.MaxCores)
+                    {
+                        // need to redistribute the number of cores
+                        var coresPerNode = Math.Max(1, _licenseStatus.MaxCores / allNodeTags.Count);
+                        foreach (var nodeTag in allNodeTags)
+                        {
+                            coresByNode[nodeTag] = coresPerNode;
+                        }
+                    }
+                }
+
+                await _serverStore.PutLicenseLimits(licenseLimits);
```

Comment:
We are known to be the leader, pretty sure we can just add that to the Raft cluster and not have to be async.